src/nl/hannahsten/texifyidea/grammar/Latex.bnf

```diff
@@ -99,27 +104,47 @@ parameter ::= optional_param | required_param {
 // will be seen as simply an open bracket, but '[x]' at the same location will
 // be parsed as optional parameter.
 // https://stackoverflow.com/a/48709143/6629569
-optional_param ::= OPEN_BRACKET optional_param_content* CLOSE_BRACKET { pin=3 }
+optional_param ::= OPEN_BRACKET ( (keyval_pair  (PARAM_SEPARATOR keyval_pair)*) | optional_param_content*) CLOSE_BRACKET { pin=3 }
 
 required_param ::= OPEN_BRACE required_param_content* CLOSE_BRACE { pin=1 }
 
 // These are like content, but no brackets and with parameter_text instead of normal_text
 // We have to separate optional and required parameter content, because required parameter content
 // can contain mismatched brackets, but optional parameters not (then we wouldn't know what to match)
-optional_param_content ::= raw_text | magic_comment | comment | environment | pseudocode_block | math_environment | COMMAND_IFNEXTCHAR | commands | group | OPEN_PAREN | CLOSE_PAREN | parameter_text
-required_param_content ::= raw_text | magic_comment | comment | environment | pseudocode_block | math_environment | COMMAND_IFNEXTCHAR | group | OPEN_PAREN | CLOSE_PAREN | parameter_text | OPEN_BRACKET | CLOSE_BRACKET
+optional_param_content ::= raw_text | magic_comment | comment | environment | pseudocode_block | math_environment | COMMAND_IFNEXTCHAR | commands | group | OPEN_PAREN | CLOSE_PAREN | parameter_text | NORMAL_TEXT_CHAR
+required_param_content ::= raw_text | magic_comment | comment | environment | pseudocode_block | math_environment | COMMAND_IFNEXTCHAR | group | OPEN_PAREN | CLOSE_PAREN | parameter_text | PARAM_SEPARATOR | KEYVAL_ASSIGN | OPEN_BRACKET | CLOSE_BRACKET | NORMAL_TEXT_CHAR
+
+keyval_pair ::= keyval_key (KEYVAL_ASSIGN keyval_value)?
+keyval_key ::= keyval_content+ {
+    methods=[toString]
+}
+keyval_value ::= keyval_content+ {
+    methods=[toString getNameIdentifier getName setName]
+}
+keyval_content ::= commands | keyval_text | parameter_group
+
+// We need to include the special characters here because changing NORMAL_TEXT_CHAR would alter the lexer's behaviour
```

Comment:
Is this comment obsolete now, because you did change normal text char, right? I would expect parameter_text in the line above, intead of keyval_text. That also saves duplicating some things that exist for parameter_text for keyval_text, like for the find usages I think (which now almost works but not quite).

Reply:
> Is this comment obsolete now, because you did change normal text char, right?

It is not, unfortunately. It is basically the same issue though. I could have removed these characters from `NORMAL_TEXT_CHAR` and created separate tokens for them. However, that would have meant to include all of them in every non-terminal that currently uses `NORMAL_TEXT_CHAR`. If you prefer that, I can change it.

> I would expect parameter_text in the line above, intead of keyval_text. That also saves duplicating some things that exist for parameter_text for keyval_text, like for the find usages I think (which now almost works but not quite).
> 
> It's a bit confusing though, I'm not sure I'm suggesting the right things so maybe I should just try myself.
> 
> To clarify:
> `\lstinputlisting[label=<caret>{lst:inputlisting}]{some/file}` does work for Find Usages, but
> `\lstinputlisting[label={lst:inp<caret>utlisting}]{some/file}` does not

I have answered this part in the comment below.